integrations: Update GoCD integration.

PieterCK opened the pull request:
This PR updates the GoCD integration to use [Sentry's WebHook notifier plugin](https://github.com/getsentry/gocd-webhook-notification-plugin/releases). 

We're using version **v0.0.6 of Sentry's WebHook notifier plugin** because the more recent version (v0.0.9) seems to fail to send requests from the gocd server machine. error logs and more detail in [CZO](https://chat.zulip.org/#narrow/stream/127-integrations/topic/GoCD.20integration.20not.20working.20.2321224/near/1932445) and in this [issue](https://github.com/getsentry/gocd-webhook-notification-plugin/issues/9) that I've opened in the plugins repo.




Old PR: #28746
Fixes: #21224
[CZO discussion](https://chat.zulip.org/#narrow/stream/127-integrations/topic/GoCD.20integration.20not.20working.20.2321224)

**Screenshots and screen captures:**
<details>

<summary>Updated doc</summary>

![image](https://github.com/user-attachments/assets/be2ff6c1-f4ff-4a12-94c0-5e77a82f91ab)

</details>


<details>

<summary>Notification screenshots</summary>

![image](https://github.com/user-attachments/assets/7d716656-7806-41a6-b477-f2a85604daad)

</details>

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->



<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

PieterCK commented:
@zulipbot add "buddy review" 

@kennethnrk Could you please review this one? Thanks!

PieterCK commented:
@kennethnrk thanks for checking this out. 

>Tested it using the integrations testing Devtools. LGTM. Do i need to setup a GoCD server to test this issue or is the devtool good enough?

I've already tested it end to end with my local gocd server, but I guess this is worth mentioning for other reviewers.

PieterCK commented:
@alya - Do you think this is ready for maintainer review? I think my PR will go through any mentor review anymore

PieterCK commented:
@zulipbot remove "buddy review"

alya commented:
Sure, @laurynmm, are you up for taking a look?

timabbott commented on zerver/webhooks/gocd/doc.md:
```diff
@@ -1,24 +1,39 @@
+!!! warn ""
+
+    **Note**: We currently only support version **v0.0.6** of Sentry's GoCD notification
+    plugin.
```
I think it might make sense to  just roll this into the "Download the far file ..." step? They just need to make sure they have version v0.0.6 or higher in that step, if it's already installed (but presumably they don't need to check anything if it's already new).

alya commented:
Do we literally support just v0.0.6 and not anything older or newer? If so, why is that?

Releases, for reference: https://github.com/getsentry/gocd-webhook-notification-plugin/releases

PieterCK commented:
I've tested running using the most recent version (v0.0.9) of the plugin but it doesn't work. I think it's the plugin because I noticed an error related to the plugin in the gocd servers log. More details here: [CZO](https://chat.zulip.org/#narrow/stream/127-integrations/topic/GoCD.20integration.20not.20working.20.2321224/near/1932445).

On the same setup, I tried using v0.0.6 which is the same version the old PR(#28746) was using and it worked. I've also filed an issue on their repo here: [issue](https://github.com/getsentry/gocd-webhook-notification-plugin/issues/9)

PieterCK commented:
Update: I've updated the doc and the PR description, hopefully it's clearer now.

PieterCK commented:
Update: Cleaned up a couple of unused screenshots & tidied up reference links in the doc

alya commented:
@laurynmm would you be up for taking a look at this one?

laurynmm commented on zerver/webhooks/gocd/view.py:
```diff
@@ -9,13 +9,26 @@
+TOPIC_TEMPLATE = "{pipeline} / {stage}"
+
+SCHEDULED_BODY_TEMPLATE = """
+**Pipeline** {pipeline}/{stage}: {status}.
```
This isn't a sentence, so let's remove the period at the end of the line. I'd also move the status information ...

`**Pipeline {status}**: {pipeline} / {stage}`

PieterCK replied:
Done.

laurynmm commented on zerver/webhooks/gocd/view.py:
```diff
@@ -9,13 +9,26 @@
-Details: [build log]({})
-Comment: {}"""
+COMMIT_INFO_TEMPLATE = (
+    """Triggered on [`{commit_details}`]({commit_link}) on branch `{branch_name}`."""
```
Looking at the message I feel like, "**Commit**: `commit-sha-link` on branch `name`" would be fine. I'm not sure what's gained by "Triggered on"? I'd also drop the period here.

PieterCK replied:
hmm yeah, that'd probably be fine.

laurynmm commented on zerver/webhooks/gocd/view.py:
```diff
@@ -9,13 +9,26 @@
+"""
+
+COMPLETED_BODY_TEMPLATE = """
+**Build** {pipeline}/{stage}: {status} {emoji}.
```
I'd lead with the emoji, put the status in the bold section ...

`{emoji} **Build {status}**: {pipeline} / {stage}`

laurynmm commented on zerver/webhooks/gocd/view.py:
```diff
@@ -26,31 +39,75 @@ def api_gocd_webhook(
-    return json_success(request)
+def get_commit_details(payload: WildValue) -> str:
+    material = payload["data"]["pipeline"]["build-cause"][0]["material"]
+    url = material["git-configuration"]["url"].tame(check_string)
```
Maybe `url_base` here so that it's clearer the `url` in the template has the commit hash?

Also, all this information comes from "build-cause", so maybe get that first and then the url, branch and revision from that?

PieterCK replied:
Doone.
```diff
def get_commit_details(payload: WildValue) -> str:
-    material = payload["data"]["pipeline"]["build-cause"][0]["material"]
-    url = material["git-configuration"]["url"].tame(check_string)
-    revision = payload["data"]["pipeline"]["build-cause"][0]["modifications"][0]["revision"].tame(
-        check_string
-    )
+    build_casue = payload["data"]["pipeline"]["build-cause"]
+    material = build_casue[0]["material"]
+    url_base = material["git-configuration"]["url"].tame(check_string)
+    revision = build_casue[0]["modifications"][0]["revision"].tame(check_string)
     commit_sha = get_short_sha(revision)
-    url = f"{url}/commit/{commit_sha}"
+    url = f"{url_base}/commit/{commit_sha}"
     branch = material["git-configuration"]["branch"].tame(check_string)
     return COMMIT_INFO_TEMPLATE.format(
         commit_details=commit_sha,
```

laurynmm commented on zerver/webhooks/gocd/view.py:
```diff
@@ -26,31 +39,75 @@ def api_gocd_webhook(
+
+    for key in formatted_job_dict:
+        formatted_job_list = ", ".join(formatted_job_dict[key])
+        job_details_template += f"- **{key} job(s)**: {formatted_job_list}\n"
```
I'd drop the "job(s)" part and just have it be the key, which seems to be "Failed" or "Passed".

laurynmm commented on zerver/webhooks/gocd/view.py:
```diff
@@ -26,31 +39,75 @@ def api_gocd_webhook(
+        "status": pipeline_data["stage"].get("state").tame(check_string),
+        "pipeline": pipeline_data["name"].tame(check_string),
+        "stage": pipeline_data["stage"]["name"].tame(check_string),
+        "start_time": pipeline_data["stage"]["create-time"].tame(check_string),
```
This would be easier for me to read if we were consistent between using `.get()` or just the  `["key-1"]["key-2"]` format here and below.

PieterCK replied:
Done
```diff
@@ -91,7 +88,7 @@ def get_body(payload: WildValue) -> str:
     pipeline_data = payload["data"]["pipeline"]
     body_details = {
         "commit_details": get_commit_details(payload),
-        "status": pipeline_data["stage"].get("state").tame(check_string),
+        "status": pipeline_data["stage"]["state"].tame(check_string),
         "pipeline": pipeline_data["name"].tame(check_string),
         "stage": pipeline_data["stage"]["name"].tame(check_string),
         "start_time": pipeline_data["stage"]["create-time"].tame(check_string),
@@ -99,12 +96,12 @@ def get_body(payload: WildValue) -> str:
     if body_details["status"] == "Building":
         return SCHEDULED_BODY_TEMPLATE.format(**body_details)
     else:
-        result = pipeline_data["stage"].get("result").tame(check_string)
+        result = pipeline_data["stage"]["result"].tame(check_string)
         body_details.update(
             {
                 "result": result,
                 "emoji": ":check:" if result == "Passed" else ":warning:",
-                "end_time": pipeline_data["stage"].get("last-transition-time").tame(check_string),
+                "end_time": pipeline_data["stage"]["last-transition-time"].tame(check_string),
             }
         )

```

laurynmm replied:
So, those keys might not be present? Then I think we need the `check_none_or` in that case, right?

PieterCK replied:
oops, I meant it's probably ok to use ["key1"]["key2"] as those keys exist in all known payload types

laurynmm commented on zerver/webhooks/gocd/doc.md:
```diff
@@ -1,24 +1,33 @@
+1. {!generate-webhook-url-basic.md!}
 
-1. Add the following to your `Config.XML` file.
+1. Download and install version **v0.0.6** of Sentry's [**WebHook Notifier    plugin**][gocd-webhook-link]. Please note that newer versions might not work
```
This line needs to be fixed to have a break instead of a tab.

laurynmm commented on zerver/webhooks/gocd/doc.md:
````diff
@@ -1,24 +1,33 @@
-        ...
-    </pipeline>
-    ```
+1. In your GoCD server, go to **Admin>Server Configuration>Plugins**. Click
````
Let's add spaces around the `>` here ... "**Admin > Server Configuration > Plugins**" for readability.

laurynmm reviewed (commented):
@PieterCK - Thanks for working on these updates! I had some feedback on the new notification message format, and noticed some issues with the documentation updates.

Let me know if you have any questions about my comments.

It looks like the issue with the newest version of the plugin is still outstanding? Did you check the intermediate versions of the plugin to try and see when the breaking changes were added?

PieterCK commented:
Thanks for reviewing this! will address them tomorrow

PieterCK commented:
> It looks like the issue with the newest version of the plugin is still outstanding? Did you check the intermediate versions of the plugin to try and see when the breaking changes were added?

Yeah, it seems like that's still the case. And no, I didn't.

PieterCK commented:
This is ready for you to take a look at again @laurynmm , I've also updated the PR description for screenshots. Let me know if any parts are unclear. Thanks again

PieterCK commented:
This is ready for you to take a look at again @laurynmm , I've also updated the PR description for screenshots. Let me know if any parts are unclear. Thanks again

PieterCK commented:
Update: cleaned up variable names, added an early return, and cleaned up the integration doc.

<details>

<summary>diff</summary>

```diff
pieter@PIETER-UBUNTU:~/projects/zulip$ git diff
diff --git a/zerver/webhooks/gocd/doc.md b/zerver/webhooks/gocd/doc.md
index 2f21d133c2..6a7d968ab3 100644
--- a/zerver/webhooks/gocd/doc.md
+++ b/zerver/webhooks/gocd/doc.md
@@ -10,9 +10,8 @@ Get GoCD notifications in Zulip!
 
 1. {!generate-webhook-url-basic.md!}
 
-1. Download and install version **v0.0.6** of Sentry's
-   [**WebHook Notifier plugin**][gocd-webhook-link]. Please note that
-   newer versions might not work properly.
+1. Download and install version **v0.0.6** of Sentry's [**WebHook Notifier
+   plugin**][1]. Please note that newer versions might not work properly.
 
 1. In your GoCD server, go to **Admin > Server Configuration > Plugins**.
    Click on the gear icon beside the **WebHook Notifier plugin** you just installed, paste the generated URL, and click **Save**.
@@ -25,9 +24,9 @@ Get GoCD notifications in Zulip!
 
 ### Related Branches
 
-- [GoCD plugins documentation][gocd-plugins-doc-link]
+- [GoCD plugins documentation][2]
 
 {!webhooks-url-specification.md!}
 
-[gocd-webhook-link]: https://github.com/getsentry/gocd-webhook-notification-plugin/releases/tag/v0.0.6
-[gocd-plugins-doc-link]: https://docs.gocd.org/current/extension_points/plugin_user_guide.html#installing-and-uninstalling-of-plugins
+[1]: https://github.com/getsentry/gocd-webhook-notification-plugin/releases/tag/v0.0.6
+[2]: https://docs.gocd.org/current/extension_points/plugin_user_guide.html#installing-and-uninstalling-of-plugins
diff --git a/zerver/webhooks/gocd/view.py b/zerver/webhooks/gocd/view.py
index 163c4d471e..d5322524ad 100644
--- a/zerver/webhooks/gocd/view.py
+++ b/zerver/webhooks/gocd/view.py
@@ -53,10 +53,10 @@ def get_topic(payload: WildValue) -> str:
 
 
 def get_commit_details(payload: WildValue) -> str:
-    build_casue = payload["data"]["pipeline"]["build-cause"]
-    material = build_casue[0]["material"]
+    build_cause = payload["data"]["pipeline"]["build-cause"]
+    material = build_cause[0]["material"]
     url_base = material["git-configuration"]["url"].tame(check_string)
-    revision = build_casue[0]["modifications"][0]["revision"].tame(check_string)
+    revision = build_cause[0]["modifications"][0]["revision"].tame(check_string)
     commit_sha = get_short_sha(revision)
     url = f"{url_base}/commit/{commit_sha}"
     branch = material["git-configuration"]["branch"].tame(check_string)
@@ -93,18 +93,19 @@ def get_body(payload: WildValue) -> str:
         "stage": pipeline_data["stage"]["name"].tame(check_string),
         "start_time": pipeline_data["stage"]["create-time"].tame(check_string),
     }
+
     if body_details["status"] == "building":
         return SCHEDULED_BODY_TEMPLATE.format(**body_details)
-    else:
-        result = pipeline_data["stage"]["result"].tame(check_string)
-        body_details.update(
-            {
-                "result": result,
-                "emoji": ":check:" if result == "Passed" else ":warning:",
-                "end_time": pipeline_data["stage"]["last-transition-time"].tame(check_string),
-            }
-        )
-        body = COMPLETED_BODY_TEMPLATE.format(**body_details)
-
-        body += get_jobs_details(pipeline_data)
-        return body
+
+    result = pipeline_data["stage"]["result"].tame(check_string)
+    body_details.update(
+        {
+            "result": result,
+            "emoji": ":check:" if result == "Passed" else ":warning:",
+            "end_time": pipeline_data["stage"]["last-transition-time"].tame(check_string),
+        }
+    )
+    body = COMPLETED_BODY_TEMPLATE.format(**body_details)
+
+    body += get_jobs_details(pipeline_data)
+    return body
```
</details>

laurynmm commented on zerver/webhooks/gocd/view.py:
```diff
@@ -26,31 +37,75 @@ def api_gocd_webhook(
 
-    return json_success(request)
+def get_commit_details(payload: WildValue) -> str:
+    build_cause = payload["data"]["pipeline"]["build-cause"]
```
Could this be `build = payload["data"]["pipeline"]["build-cause"][0]` and then you don't need to get the material variable, but can use that to get the branch, base URL and commit hash?

laurynmm commented on zerver/webhooks/gocd/view.py:
```diff
@@ -26,31 +37,75 @@ def api_gocd_webhook(
+    body_details.update(
+        {
+            "result": result,
+            "emoji": ":check:" if result == "Passed" else ":warning:",
```
Based on [this CZO discussion](https://chat.zulip.org/#narrow/channel/137-feedback/topic/markdown.20alerts.20syntax/near/1961518), should we maybe use `":green_circle:"` and `":red_circle:"` here? Changing this would mean updating the generated screenshot. This isn't an alert/monitoring integration, but it might be good to align with that general direction for now.

laurynmm reviewed (commented):
@PieterCK - This looks great! Had two small thoughts, but otherwise this looks almost good to go.

Any word from the plug-in about the issue you reported? Were you able to try the different versions between the current v0.0.9 release and v0.0.6 to see if you could pinpoint when the breaking change was introduced?

laurynmm commented:
@PieterCK - I pushed a few small revisions to the view code here; see updates below.

@timabbott - I think the updates here would be good for integration review except for the outstanding issue with the upstream plugin. Thoughts on the next steps for that?

**Updates**:
- Simplified the `get_commit_details` function a bit.
- Adjusted some spacing so the templates are easier to read.

PieterCK commented:
Turns out the newer version can't send requests to a local server, so after trying it on CZO using a JSON incoming webhook bot I was able to get it to work.

The payload from the newer plugin version is the same, so I only updated the documentation (PR description).

laurynmm commented:
I think the changes to the plug-in is that it no longer works for HTTP and only works for HTTPS, so we should note that vs it not working in the development environment. The Stripe integration has a note about only sending payloads over HTTPS that you could maybe use as an example.

PieterCK commented:
Ah, Ok. I've updated the docs!

laurynmm commented:
Made a few revisions to the documentation as I didn't like that note at the top in this case. See updated screenshot below.

Marking this for integration review.

<details>
<summary>Updated GoCD integration doc</summary>

![Screenshot from 2024-11-07 16-04-04](https://github.com/user-attachments/assets/5853766c-62a3-488f-96e1-dc886bb970fd)
</details>

timabbott commented:
Merged, thanks @PieterCK and @laurynmm!

swayam0322 commented:
Damn feels like yesterday when I started working on this
